.github/scripts/toolchain/group-gcc-results-by-source.py

```diff
@@ -0,0 +1,115 @@
+                        break
+
+    print(f"Finished reading all log files.")
+    print()
+    
+    total_test_count = sum(len(messages) for messages in log_messages.values())
+    print(f"Found logs from {total_test_count} tests. Number of tests per result type:")
+    for log_type, messages in log_messages.items():
+        print(f"  {log_type: <12} {len(messages)}")
+    print()
+
+    return log_messages
+
+
+def print_path_structure(log_messages: LOG_MESSAGES_TYPE) -> None:
+    def parse_filepath(message: list[str]) -> list[str]:
+        # We parse here the last line of each log message, example:
+        # FAIL: g++.dg/contracts/contracts-pre10.C   20 blank line(s) in output"""
+        assert any(message[-1].startswith(log_type) for log_type in LOG_TYPES)
+        filepath = message[-1].split(" ")[1]
+        # The path might or but don't have to start with tool name split by dot
+        if "." in filepath and "/" not in filepath.split(".", 1)[0]:
+            tool_name, filepath = filepath.split(".", 1)
+            return [tool_name] + filepath.split("/")
+        return filepath.split("/")
+    
+    @dataclasses.dataclass
+    class PathSegmentStats:
+        count: int = 0
+        children: dict[str, "PathSegmentStats"] = dataclasses.field(default_factory=dict)
+
+    def print_path_segment_stats(entry: PathSegmentStats, indent: int = 0, print_limit: int = 100) -> None:
+        """Prints hierarchical path segments with their occurrence counts
+        
+        print_limit: Print only path segments with occurrence count >= print_limit
+        """
+        sorted_path_stats = sorted(entry.children.items(), key=lambda x: x[1].count, reverse=True)
+        for key, value in sorted_path_stats:
+            if value.count < print_limit:
+                print(" " * (indent + 2) + "...")
+                return
+            
+            print(" " * indent + f"{key}: {value.count}")
+            print_path_segment_stats(value, indent + 2)
+
+    root_seg_stats = PathSegmentStats()
+    for messages in log_messages.values():
+        for message in messages:
+            path_segments = parse_filepath(message)
+            curr_seg_stats = root_seg_stats
+            curr_seg_stats.count += 1
+            for segment in path_segments:
+                if segment not in curr_seg_stats.children:
+                    curr_seg_stats.children[segment] = PathSegmentStats()
+                curr_seg_stats = curr_seg_stats.children[segment]
+                curr_seg_stats.count += 1
+
+    print_path_segment_stats(root_seg_stats)
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(description="Aggregate test results from log files")
+    parser.add_argument('DIR', type=str, help="Directory containing log files")
+    args = parser.parse_args()
+
+    log_messages = read_logs(args.DIR)
+    # We don't care about tests that are passing, remove them
+    for log_type in PASS_LOG_TYPES:
+        log_messages.pop(log_type)
```

Comment:
Wouln't be better to skip collection of information about passes at all? Do you need information about passes for something?

Reply:
True, my main motivation was debugging = to have the output to see that the script loaded all the messages. Agree it don't have to be in the final script since it's already in summary